[Routing] dump static arrays instead of classes for both matcher and generator

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Instead of dumping a class for the Url generator and matcher, I'd like to try dumping a static array, and use a generic class to walk through the array.

This should allow leveraging the PHP7 cache for static arrays, thus make the router matcher scale to a big number of routes without much performance penalty.

I'd happily accept help for benching this on your apps.

nicolas-grekas committed:
[Routing] dump static arrays instead of classes for both matcher and generator

iltar commented on src/Symfony/Component/Routing/Tests/Fixtures/dumper/static_url_matcher3.php:
```diff
@@ -0,0 +1,57 @@
+                    [
+                        [
+                            'match',
+                            '#^/rootprefix/(?P<var>[^/]++)$#s',
```
Maybe not entirely related to this PR, but would it also be an idea to optimize routes away that can't ever be matched? For example, I register a bunch of routes which are exposed via a bundle, but they link to an external application. Those routes will show up in the URL matcher, while I really only need them in the generator.

Example:
```
// frontend_routing.hostnetnl.search
if ('/zoeken' === $pathinfo) {
    return $this->mergeDefaults(array_replace($hostMatches, array('_route' => 'frontend_routing.hostnetnl.search')), array ());
}
```

The reason I mention this, is because while being a URL matcher fixture, it doesn't seem like an actual fixture that could be called if used in a real world application. Currently not too much overhead because they are loaded last in my scenario, but it's still 51 routes in the current version of my bundle. This being added to an array, but _never_ used, it feels like an optimization would be very welcome!

nicolas-grekas replied:
@iltar we should certainly not do that in this PR as this is a separate concern. The current PR is big enough :)

javiereguiluz commented on src/Symfony/Component/Routing/Matcher/Dumper/StaticUrlMatcherDumper.php:
```diff
@@ -0,0 +1,256 @@
+        $regex = $compiledRoute->getRegex();
+
+        if (!count($compiledRoute->getPathVariables()) && false !== preg_match('#^(.)\^(?P<url>.*?)\$\1#'.('u' === substr($regex, -1) ? 'u' : ''), $regex, $m)) {
+            if ($supportsTrailingSlash && '/' === substr($m['url'], -1)) {
```
Thanks to PHP 7.1 we could remove a function call here:

```php
substr($m['url'], -1)
$m['url'][-1]
```

javiereguiluz commented on src/Symfony/Component/Routing/Matcher/StaticUrlMatcher.php:
```diff
@@ -0,0 +1,139 @@
+                        $ret = $defaults;
+                    }
+
+                    if ($checkTrailingSlash && substr($pathinfo, -1) !== '/') {
```
Here too:

```php
substr($pathinfo, -1) !== '/'
'/' !== $pathinfo[-1]
```

javiereguiluz commented on src/Symfony/Component/Routing/Tests/Fixtures/dumper/static_url_matcher5.php:
```diff
@@ -0,0 +1,199 @@
+<?php
```
I appreciate the efforts to generate a somewhat readable matcher files, but looking at the generated files (like this one) I wonder if it's readable at all with the route definition spanning so many lines.

nicolas-grekas replied:
should be much better now

javiereguiluz replied:
It's great now. Thank you!

nicolas-grekas commented:
PR ready on my side. A one-route app show no measurable performance impact. I'd be happy to gather performance number on real apps with bigger routing table.

Maybe @frankdejonge since you already worked on that topic? Anyone else really also :)

nicolas-grekas commented on .php_cs.dist:
```diff
@@ -40,5 +40,7 @@ return PhpCsFixer\Config::create()
             ->notPath('Symfony/Bridge/PhpUnit/Tests/DeprecationErrorHandler/weak.phpt')
             ->notPath('Symfony/Component/Debug/Tests/DebugClassLoaderTest.php')
+            // file autogenerated
+            ->notPath('Symfony/Component/Routing/Tests/Fixtures/dumper/static_url_matcher3.php')
```
dunno why this is ignored by fabbot, which raises a false positive for now

nicolas-grekas commented on src/Symfony/Component/Routing/Matcher/Dumper/StaticUrlMatcherDumper.php:
```diff
@@ -0,0 +1,296 @@
+        $this->expressionLanguageProviders[] = $provider;
+    }
+
+    private function compileRoutes(RouteCollection $routes): array
```
@alexpott you might be interested into having this method public at some point for Drupal, as it allows building a static routing table that can be (json)serialized and used together with `StaticUrlMatcher`

dawehner replied:
Right now our entire routing system uses a database tables. This was introduced due to some reasons:

* There might be a lot of possible routes
* User changes might trigger new routes

We then leverage the DB to do our filtering, instead of something purely regex based. Does that make sense for you?

We explored a bit the idea of doing some static optimizations for like the frontpage or so, but quickly realized that there is too much flexibility in the existing system and more important, the bottleneck is not the routing system at all.

nicolas-grekas replied:
yep, makes sense, yet I don't know if it's actually better to model routes in the DB (some rows per routes, which is what you do now?), vs just putting one json that describes the full routing in the DB, which could be generated here. I don't have the answer, I just wanted to let you know :)

frankdejonge commented:
@nicolas-grekas I've anonymised the routes dump from the reference project I used. You can use this as a reference. I parsed it before to create the route collections: https://gist.github.com/frankdejonge/65e92a4bfa9e0daed6974e0a82ad0684

stof commented on src/Symfony/Component/Routing/Matcher/StaticUrlMatcher.php:
```diff
@@ -0,0 +1,149 @@
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+abstract class StaticUrlMatcher extends UrlMatcher implements RedirectableUrlMatcherInterface
```
Are we required to make this an abstract class forcing the child to implement RedirectableUrlMatcherInterface ? The redirectable part is the only un-implemented part.
This PR makes the Routing component unusable in a dumped context without using FrameworkBundle (or implementing your own class).

In the existing matcher, support for redirecting is optional. We could do the same here by removing this abstract interface here (making the class concreate) and adding a `if ($this instanceof RedirectableUrlMatcherInterface)` in places needing to change the logic for redirections ? What would be the performance impact of this check for the framework case where it is redirectable ?

stof commented on src/Symfony/Component/Routing/Matcher/StaticUrlMatcher.php:
```diff
@@ -0,0 +1,149 @@
+                                if (0 !== strpos($pathinfo, $conditions[1])) {
+                                    continue 2;
+                                }
+                                // no break
```
instead of doing a fallthrough and then changing the matched conditions by changing the index with an int cast of the type check, it would be much more readable to duplicate the `preg_match` call, passing the right condition in each case.

nicolas-grekas replied:
updated

stof commented:
I would like to see benchmarks for both generating and matching here, as well as extensive testing that the new matcher behaves the same than the old one (as the logic is totally rewritten).

frankdejonge commented:
@stof out of curiosity, why would you want benchmarks for generating? The generation of the routes should be done before something is deployed and shipped in the artefact/container/deployable-zip/whatever. At least, that's my view on the subject, curious to hear what reasons you'd have to measure the performance of this.

javiereguiluz commented:
@frankdejonge I think Christophe refers to generate a URL from a route (`{{ path('my_route')`) instead of generating these static files.

frankdejonge commented:
Ahhhhh that makes sense.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Tests/Routing/StaticUrlMatcherTest.php:
```diff
@@ -0,0 +1,77 @@
+    private function getMatcher(RouteCollection $routes, RequestContext $context)
+    {
+        $dumper = new StaticUrlMatcherDumper($routes);
+        $path = sys_get_temp_dir().DIRECTORY_SEPARATOR.'php_matcher.'.uniqid('StaticUrlMatcher').'.php';
```
fixed (this was copy/paste from existing code)

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Routing/StaticUrlMatcher.php:
```diff
@@ -0,0 +1,42 @@
+/**
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+class StaticUrlMatcher extends BaseMatcher
```
See reply in comments.

frankdejonge commented:
@nicolas-grekas did you see the gist with the routes?

nicolas-grekas commented:
> Are we required to make this an abstract class forcing the child to implement RedirectableUrlMatcherInterface ?

I don't know. Isn't this requirement legitimate? Is it a hard one to fulfill? It feels like all router should have this redirect method to me. Technically, the dumped code is currently dependent whether `RedirectableUrlMatcherInterface` is implemented by the matcher. This is something the new dumper shouldn't care about, as it's just dumping data. But dumping data that could describe the routing mapping in both situations at once (which is a requirement) is going to create overhead in the arrays, for something that is barely legitimate. So yes, just implement that method, not a serious deal.

> This PR makes the Routing component unusable in a dumped context without using FrameworkBundle (or implementing your own class).

That's already 100% the case when using dumped routes and handling redirects (which IMHO everyone does, the trailing slash issue is not specific to Symfony but a cross-tech issue with HTTP routing).

> I would like to see benchmarks for both generating and matching here, as well as extensive testing that the new matcher behaves the same than the old one (as the logic is totally rewritten).

The code for the generator is *exactly* the same - only the initialization logic has changed - so I'm not going to spend much time on it.
For the matcher, I agree we're missing a test suite, too bad we didn't have it already.

stof commented:
> The code for the generator is exactly the same - only the initialization logic has changed - so I'm not going to spend much time on it.

It would still make sense to have a benchmark to see whether we actually benefit from the static array caching.

> For the matcher, I agree we're missing a test suite, too bad we didn't have it already.

It might make sense to work on building such testsuite first in a separate PR (which means it would run against the current implementation) and then rebase this PR on top of it (ensuring that the new implementation passes it too)

stof commented:
@frankdejonge your dump does not allow rebuilding the routes fully (requirements are not available in it for instance)

frankdejonge commented:
@stof in my case I just used a wildcard matcher because the changes didn't happen on that front. Doesn't seem to be a factor here either. The difference is the lookup mechanism, not those checks.

stof commented:
@frankdejonge as the matching logic has been fully rewritten, I would like to see benchmarks for cases using all features too, not only for a single case.

frankdejonge commented:
I've deleted the route dump. Have fun building feature complete fixtures 👍

nicolas-grekas commented:
> I've deleted the route dump. Have fun building feature complete fixtures

don't! it's no fun :)

> Are we required to make this an abstract class forcing the child to implement RedirectableUrlMatcherInterface ?

Fixed - was easier than anticipated, this adds no overhead.

> It would still make sense to have a benchmark to see whether we actually benefit from the static array caching.

the current code already benefits from static array caching, since it already dumps a static array - this is really 100% the same.

fancyweb commented:
I did some benchmarks using one of my project routes (789 routes).
1000 revs, 3 iterations.
Winner is in bold.

`PhpMatcherDumper` vs **StaticUrlMatcherDumper** : 

benchmark | subject | groups | params | revs | iter | mem_peak | time_rev | comp_z_value | comp_deviation
 --- | --- | --- | --- | --- | --- | --- | --- | --- | --- 
MatcherDumperBench | benchPhpMatcherDumper |  | [] | 1000 | 0 | 122,099,944b | 13,925.855μs | +1.37σ | +0.87%
MatcherDumperBench | benchPhpMatcherDumper |  | [] | 1000 | 1 | 122,099,944b | 13,719.996μs | -0.98σ | -0.62%
MatcherDumperBench | benchPhpMatcherDumper |  | [] | 1000 | 2 | 122,099,944b | 13,771.434μs | -0.39σ | -0.25%
MatcherDumperBench | benchStaticUrlMatcherDumper |  | [] | 1000 | 0 | 122,238,000b | 12,447.356μs | -1.2σ | -0.61%
MatcherDumperBench | benchStaticUrlMatcherDumper |  | [] | 1000 | 1 | 122,238,000b | 12,603.555μs | +1.25σ | +0.63%
MatcherDumperBench | benchStaticUrlMatcherDumper |  | [] | 1000 | 2 | 122,238,000b | 12,521.185μs | -0.04σ | -0.02%


**PhpGeneratorDumper** vs `StaticUrlGeneratorDumper` : 

benchmark | subject | groups | params | revs | iter | mem_peak | time_rev | comp_z_value | comp_deviation
 --- | --- | --- | --- | --- | --- | --- | --- | --- | --- 
GeneratorDumperBench | benchPhpGeneratorDumper |  | [] | 1000 | 0 | 11,868,672b | 4,727.980μs | +0.55σ | +0.16%
GeneratorDumperBench | benchPhpGeneratorDumper |  | [] | 1000 | 1 | 11,868,672b | 4,732.192μs | +0.85σ | +0.25%
GeneratorDumperBench | benchPhpGeneratorDumper |  | [] | 1000 | 2 | 11,868,672b | 4,700.886μs | -1.4σ | -0.41%
GeneratorDumperBench | benchStaticUrlGeneratorDumper |  | [] | 1000 | 0 | 11,439,640b | 5,867.291μs | +1.40σ | +0.09%
GeneratorDumperBench | benchStaticUrlGeneratorDumper |  | [] | 1000 | 1 | 11,439,640b | 5,858.939μs | -0.87σ | -0.05%
GeneratorDumperBench | benchStaticUrlGeneratorDumper |  | [] | 1000 | 2 | 11,439,640b | 5,860.187μs | -0.53σ | -0.03%


`UrlGenerator` vs **StaticUrlGenerator** (each route is generated 10 times) : 

benchmark | subject | groups | params | revs | iter | mem_peak | time_rev | comp_z_value | comp_deviation
 --- | --- | --- | --- | --- | --- | --- | --- | --- | --- 
GeneratorBench | benchUrlGenerator |  | [] | 1000 | 0 | 16,935,960b | 26,957.715μs | +1.18σ | +2.76%
GeneratorBench | benchUrlGenerator |  | [] | 1000 | 1 | 16,935,960b | 25,458.014μs | -1.27σ | -2.96%
GeneratorBench | benchUrlGenerator |  | [] | 1000 | 2 | 16,935,960b | 26,287.435μs | +0.09σ | +0.20%
GeneratorBench | benchStaticUrlGenerator |  | [] | 1000 | 0 | 16,935,960b | 25,434.804μs | +1.10σ | +2.79%
GeneratorBench | benchStaticUrlGenerator |  | [] | 1000 | 1 | 16,935,960b | 23,923.030μs | -1.32σ | -3.32%
GeneratorBench | benchStaticUrlGenerator |  | [] | 1000 | 2 | 16,935,960b | 24,877.948μs | +0.21σ | +0.54%

stof commented:
@fancyweb benchmarking a dumper vs the other is irrelevant. Dumping happens during cache warmup, where we can spend time.

What we need is to compare the matching using the dumped matcher vs the matching using the static matcher with dumped routes. And this is what you don't have in your comment.

nicolas-grekas commented:
@fancyweb: @stof is right, sorry if that's wasn't clear enough. The perf comparison should be about both url matchers: the one generated by PhpDumperMatcher, vs a StaticUrlMatcher one.

Since bootstrap time should be considered, I think the way to bench this is taking a skeleton, load it with some fixtures routes, and measure the response time before/after this PR.

nicolas-grekas commented:
Matcher now tested, with the existing test cases from UrlMatcher.

nicolas-grekas commented:
As far as I tested correctly, on PHP 7.2, this is slower by a very tiny margin. The reason is easy: the matching loop has to do extra "ifs". But the difference does not diverge and is barely measurable in practice, both on small and big tables. Which means I did not hit the theoretical scalability issue with the current code. One might want to try with bigger tables (mine had ~1000 routes).

From design pov, I prefer a generic algorithm with data driven input than a dumped code. So I'm still on the side that this should be seriously considered. E.g. by looking at the data, we can already see that there are more potential optimizations that could be driven by creating hash-maps (a lot of routes could be just put into hashes and don't require the "strpos" check beforehand. Doing so at the code generation level is possible, eg using a switch on PHP 7.2, or inline hashes. But working at the data level is much more comprehensible IMHO.

The last benefit of this approach is the name of the router, which is stable. This can also be achieved in the code-generation approach, by dumping an anonymous class, now that we can.

iltar commented:
I feel like this new method allows optimizations in a way the old one did not. Instead of deprecating the current, can we simply add this new compilation under a feature flag (for now), and try to optimize it before 4.1 is released? If it turns out we can't get it faster, we can always drop it before 4.1.

stof commented:
> If it turns out we can't get it faster, we can always drop it before 4.1.

that's true even if we keep the existing approach. Until the RC release, we are free to drop this new class and undeprecate the old one if we decide to do it.

stof commented:
> we can already see that there are more potential optimizations that could be driven by creating hash-maps (a lot of routes could be just put into hashes and don't require the "strpos" check beforehand. Doing so at the code generation level is possible, eg using a switch on PHP 7.2, or inline hashes. But working at the data level is much more comprehensible IMHO.

I'm not sure this holds true considering that the order of routes is important (as soon as a route has a dynamic part, it limits optimizations for any subsequent routes). And even if we find a way to rearrange routes to optimize them (figuring out that we can switch routes because their matching is totally exclusive and so the order does not matter), such rearranging would then be usable with the dumped class too, not only with the dumped array.
This makes me think that working on re-arranging routes to unlock more optimizations might be more useful than a rewrite of battle-tested working code.

And in practice, my project has 589 routes (including the WebProfiler ones as this is in dev), among which 439 have placeholders (and so cannot be catched by a map). This lives 150 routes which could be matched by a hashmap lookup (and I think a re-arranging can move them all at the beginning given my URL convention and the fact that my requirements and route order are strict enough to avoid any ambiguity). So the re-arranging might be worth it (it is required for any useful hashmap lookup, as these 150 routes are spread over the full list, with dynamic routes appearing early (the `_wdt` one is even first).

For people wanting to perform a similar analysis, here is what I used as commands:

- counting routes: `bin/console debug:router | grep ANY |wc` if you don't use the scheme requirement or the host requirement (and so all routes have `ANY` for them). Otherwise, run `bin/console debug:router |wc` and substract 5 for the output lines not being routes (table header and table delimiters)
- counting dymaic routes: `bin/console debug:router | grep '{' |wc`

> As far as I tested correctly, on PHP 7.2, this is slower by a very tiny margin. The reason is easy: the matching loop has to do extra "ifs". But the difference does not diverge and is barely measurable in practice, both on small and big tables. Which means I did not hit the theoretical scalability issue with the current code. One might want to try with bigger tables (mine had ~1000 routes).

This makes me think that your promise (in the PR description) of making the router scaling to a higher number of routes is actually not true, as there is no performance difference.

iltar commented:
I know for a thing that sorting routes will mess up a lot, but if it means performance, I'm sure I could change my routes to not be messed up. However, there will always be some fallback routes (like wildcard `/.*`) that need to be done last. 

It almost sounds like a proper tree should be build and visualized so developers can detect url problems and probably optimize by using certain urls.

nicolas-grekas commented:
I'm closing because this demonstrated to me that this cannot be faster: traversing the static array is always going to be slower than loading + traversing the opcodes.
The implementation is finished so that if someone wants to borrow it in situations where dumping code is not possible, it's still possible.

stof commented:
@iltar this is why I'm talking about reordering routes as an optimization *when we can be sure they are exclusive* (and so their order does not matter).

And when working on implementing such logic, building a debugging tool visualizing it might indeed make sense.
